hibernate-core/src/main/antlr/org/hibernate/grammars/hql/HqlParser.g4

```diff
@@ -78,10 +78,14 @@ queryExpression
 	: queryGroup queryOrder?
 	;
 
+nestedQueryExpression
+	: LEFT_PAREN queryGroup queryOrder? RIGHT_PAREN
+	;
+
 queryGroup
-	: querySpec                                            										# QuerySpecQueryGroup
-	| LEFT_PAREN queryGroup RIGHT_PAREN															# NestedQueryGroup
-	| queryGroup queryOrder? (setOperator (querySpec | LEFT_PAREN queryGroup RIGHT_PAREN))+		# SetQueryGroup
+	: querySpec                                            						# QuerySpecQueryGroup
+	| nestedQueryExpression														# NestedQueryGroup
+	| queryGroup queryOrder? (setOperator (querySpec | nestedQueryExpression))+	# SetQueryGroup
```

Comment:
Ok that's better. so now that I understand what's going on here, I wonder if why it's ambiguous without the parents around the one with the order by.

Got an example of a query which would be ambiguous?

Reply:
The last order by clause should always refer to the whole query group i.e. `select 1 union all select 1 order by 1` should result in `QueryGroup({QuerySpec, QuerySpec}, orderBy = {1})` and if you want to order only a query spec, you need parenthesis.

Reply:
Well is that what the ANSI SQL spec says?

Because that seems like a rubbish grammar to me.

It would be better to make you write 

    (select 1 union all select 1) order by 1

or even 

    select * from (select 1 union all select 1) order by 1

fo that case.

Because all the other clauses bind to the most recent `select`. It's not normal that there should be a special use for `order by`.

Reply:
I don't have a SQL standard other than the one from 92 which was published publicly. I used the Oracle and PostgreSQL, syntax which also seems supported for TSQL:

* https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/The-UNION-ALL-INTERSECT-MINUS-Operators.html#GUID-B64FE747-586E-4513-945F-80CB197125EE
* https://www.postgresql.org/docs/current/queries-order.html
* https://github.com/antlr/grammars-v4/blob/master/sql/tsql/TSqlParser.g4#L3002

I agree that wrapping the query expression is "cleaner" and I want to support that eventually. But right now, we don't allow query expressions yet in the FROM clause

Reply:
Right, but we need to anticipate future evolution of the grammar, and not break people's queries.

why not just support the syntax:

    from (select 1 union all select 1) order by 1

as a special case for now (just especially) for this thing.

Reply:
What is the specific part that you don't like about this change? The parenthesis are useful for other things as well like operation ordering. A query like `... intersect ... union all ...` is not the same as `... intersect (... union all ...)` because set operations are executed from left to right.

Reply:
`order by` should, by nature, have a higher precedence than `union` and `intersect`, no?

I mean, every other clause of a `select` query does, right?

It's an issue of grammar design. When I originally saw that rule, I just new that something was screwy, it smells.

Reply:
We never want to say "oh this construct has such-and-such relative precedence _except_ in this special context". Or "oh in this special case you need parens when you would not expect it".

(That's the Perl way.)

Reply:
I'm pretty sure the SQL standard says that order by applies to the query group/block as all databases implement it this way. The reasoning for this is readability in multiple documentations. Here is a reference to the syntax in Oracle: https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/About-Queries-and-Subqueries.html#GUID-DB7521FE-9329-415E-B583-EA4467E990A7

In a query like `select 1 union all select 2 order by 1 desc` the order by clause does not refer to the subquery `select 2` but instead refers to the whole query group/block.

Reply:
Perhaps that's why I hate locals so much, come to think of it: they're a missed opportunity to make a new function.

Reply:
Well even if that's right, it's an objectively bad syntax, and I don't think we should copy it.

Reply:
I mean, consider:

    select foo, sum(bar) from Foo1 union select foo, sum(bar) from Foo2 group by foo order by foo

do you *really* think it's a good idea that `group by` refers to the second `select` but `order by` refers to the whole union?

That's just objectively bad.

*Either* you require people to always write

    (select foo, sum(bar) from Foo1) union (select foo, sum(bar) from Foo2 group by foo) order by foo

i.e. *always* put parens around subqueries in unions, 

*or* you require them to write

    from (select foo, sum(bar) from Foo1 union select foo, sum(bar) from Foo2 group by foo) order by foo

when they want to order the whole union.

Reply:
I have only access to SQL:2003 BNF, but ordering and limiting in subqueries was only introduced in SQL:2008 so I can't give you a definitive reference to the SQL Standard for this. Like I wrote before, all databases implement it this way (ordering applies to query group) and I bet the reason for that is that the SQL Standard mandates this.

I understand your point and I know it might not be 100% clear to a reader that doesn't know how this works. I also agree that we should (in addition) support and advocate the `select * from (...) order by ...` syntax, but before doing that we need to discuss support for query expressions in the from clause. Regardless of all of that, I'm pretty sure the SQL standard specifies that the ordering applies to the whole query group which is why I would like to match this behavior in HQL. As a user, I would find it confusing if HQL went a different route than SQL on this regard. To avoid this confusion, we'd have to prohibit ordering for set operation query groups in general and force the user to wrap the query group in `select * from (...)` or `from (...)` for the sake of adding ordering.

I'm already seeing a future user complaining and asking why Hibernate has to deviate from what the SQL Standard specifies in this regard. I'd prefer to be as close to the SQL Standard as possible and this behavior, although not immediately clear to everyone, is clearly specified.

So where do we go from here? I think that the argument "match what the SQL Standard specifies" should be enough to resolve this discussion, but if feel strongly about this, maybe this should be a team discussion/decision?

Reply:
>  I can't give you a definitive reference to the SQL Standard for this

Anyway it's beside the point, HQL isn't SQL, and we don't need to slavishly copy SQL's mistakes.

> before doing that we need to discuss support for query expressions in the from clause

Really, I don't see why. I think the correct syntax is from `from`-clause subqueries is pretty obvious, and I don't think there's going to be any debate about it at all. We know what a `from`-clause subquery is going to look like, and we have for decades&mdash;we just didn't implement it yet, because it's not so terribly useful in the context of ORM.

> As a user, I would find it confusing if HQL went a different route than SQL

> As a user, I would find it confusing if HQL went a different route than SQL on this regard.

As a user I find it confusing when languages have confusing semantics.

We should definitely not do objectively bad things just because other people did objectively bad things before us.

Ask yourself: would you be better off today if Java had reproduced more confusing things from the C/C++ languages? Was it good, for example, that Java copied the fall-through semantics of C's `switch`?

> I'm already seeing a future user complaining and asking why Hibernate has to deviate from what the SQL Standard specifies in this regard. 

And I can see _many_ more users, Java developers, who don't know SQL inside-out, being as surprised as I was that `order by` has a different precedence to `group by`.

> So where do we go from here? 

If you really think there's need for further discussion of this issue (I don't, because I know I'm right about this), then the right thing to do is to just drop this feature from this pull request, and get the other stuff that we agree on integrated.

(That's the problem with huge multi-feature pull requests.)

Reply:
> (That's the problem with huge multi-feature pull requests.)

It was actually just *luck* that I caught this before it got pushed into the language. You didn't even mention it in the list of changes!

Reply:
> Anyway it's beside the point, HQL isn't SQL, and we don't need to slavishly copy SQL's mistakes.

I don't know if it is a mistake or a deliberate decision, but the SQL standard probably also won't tell us that. In general I  agree that we shouldn't copy mistakes, but I guess this particular case is more complex as it involves not only "taste" but also the question if we want to align with what is known (SQL).

> Really, I don't see why. I think the correct syntax for from from-clause subqueries is pretty obvious, and I don't think there's going to be any debate about it at all. We know what a from-clause subquery is going to look like, and we have for decades—we just didn't implement it yet, because it's not so terribly useful in the context of ORM.

I've seen plenty of requests for it from various users, so I think subqueries in the from clause are very useful. The problem is not necessarily the HQL syntax, but more about how we represent the "type" in the query model.

> Ask yourself: would you be better off today if Java had reproduced more confusing things from the C/C++ languages? Was it good, for example, that Java copied the fall-through semantics of C's switch?

Like I wrote before, we should probably have a discussion about this and decide within the team as there are more aspects to this IMO.

> If you really think there's need for further discussion of this issue (I don't, because I know I'm right about this), then the right thing to do is to just drop this feature from this pull request, and get the other stuff that we agree on integrated.
> (That's the problem with huge multi-feature pull requests.)

I know this PR is too big, and we discussed this on Zulip already. It's not like I wanted this to be this way. I was just trying to fix testsuite issues and one thing lead to another. Future PRs will be more focused and have a corresponding issue.

> It was actually just luck that I caught this before it got pushed into the language. You didn't even mention it in the list of changes!

The feature (ordering of query groups) was there already before this PR. The changes to the grammar I did were to add support for ordering nested query groups as I apparently didn't add that originally.